File: problem-3/c-cpp/problem-3.c
#include <stdio.h>
#include "../../utils/c/primes.c"

int main()
{
    const long int given_composite = 600851475143L;
    const int composite_sqrt = 775147;

    // Create prime sieve
    unsigned char* prime_sieve = getPrimeSieve(composite_sqrt);

    // Iterate sieve in reverse and mod the given composite until it equals 0
    for (long int j = composite_sqrt - 1; j > 0; j--)
    {
        if (prime_sieve[j] != 0 && given_composite % j == 0)
        {
            printf("%ld\n", j);
            break;
        }
    }

    free(prime_sieve);

    return 0;
}

File: src/com/github/problem3/problem3.java
package com.github.problem3;

import com.github.utils.PrimeSieve;

public class problem3 {
    public static void main(String[] args) {
        long given_composite = 600851475143L;
        int composite_sqrt = (int) Math.sqrt(given_composite);

        // Create prime sieve
        boolean[] prime_sieve = PrimeSieve.NewPrimeSieve(composite_sqrt);

        // Iterate sieve in reverse and mod the given composite until it equals 0
        for (int j = composite_sqrt -1; j > 0; j--) {
            if (prime_sieve[j] && given_composite % j == 0) {
                System.out.println(j);
                break;
            }
        }
    }
}
